docs/features/software-catalog/descriptor-format.md

```diff
@@ -519,6 +519,15 @@ consumed by the component, e.g. `artist-api`. This field is optional.
 | --------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------------- |
 | [`API`](#kind-api) (default)            | Same as this entity, typically `default`   | [`consumesApi`, and reverse `apiConsumedBy`](well-known-relations.md#consumesapi-and-apiconsumedby) |
 
+### `spec.dependsOn` [optional]
+
+An array of [entity references](#string-references) to the Resources that the
+component depends on, e.g. `artists-db`. This field is optional.
+
+| [`kind`](#apiversion-and-kind-required) | Default [`namespace`](#namespace-optional) | Generated [relation](well-known-relations.md) type                                            |
+| --------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------- |
+| [`Resource`](#kind-resource) (default)  | Same as this entity, typically `default`   | [`dependsOn`, and reverse `dependencyOf`](well-known-relations.md#dependson-and-dependencyof) |
```

Comment:
This is correct in terms of the diagram - but I'm thinking that we should permit dependencies at least on `Component` kinds too.

On the one hand, we could postpone that and get this great change in without blocking too much.

On the other hand, the main risk with that is that we might want to NOT have a default here if there's no natural best default. And removing the default would be a breaking change, making people's YAML files break.

So ...

Maybe best to already now make both `Component` and `Resource` supported, and have no default?

@backstage/maintainers @Fox32 @andrewthauer @adamdmharvey @dtuite @dhenneke as people of great wisdom and customer insight - how do you feel about `Component dependsOn Component`? There's more stuff than just APIs that one can need for one's functioning - are all those other stuff `Resource`s?

Reply:
`Component dependsOn Component` is something I would like to see in the long term. Could be suited to show dependencies between libraries/packages. We could parse an NPM registry/maven repo/you name it and show which components depend on each other. I wouldn't implement it now though.

If I think about the default, I can't decide whether `Component` or `Resource` should be the default. So let's keep it open for now. Adding a default later would be a non breaking change I guess.

Reply:
Agreed. So the vote is to not intentionally limit the kinds here, and also to have no default right

Reply:
I'm in full agreement of supporting `Component dependsOn Component` as well. In some cases it might be more common to use that relation then `Resource` since a component would likely be something you own and maintain vs a resource is typically an external concern (e.g. cloud service).

> We could parse an NPM registry/maven repo/you name it and show which components depend on each other. I wouldn't implement it now though.

We've thought about doing this as well and agree it would be very useful. Just thinking out loud but ... how would this work for 3rd party libraries? Would those also be Components that are depended on by an internal Component? Or possibly a relation to a new entity type that is not necessarily tracked directly in the catalog. Maybe something else.

Reply:
Yeah i've been pondering concerns like that as well. To express relations to things, we now _have_ to import them. What if the catalog rather could express relations and metadata on top of essentially "foreign keys" - so that you could for example say that you depend on `<URN or URL expressing a specific NPM package>`, instead of just entity refs? I dunno, it's just a flicker of an idea in my mind.

Reply:
What about taking the approach of using a separate field for each of these types?  Perhaps "dependsOnResources" and "dependsOnComponents" or some similar type of set of fields to differentiate this?  This would be similar to the approach used by the `Group` entity when specifying child groups and direct members (requested in #5072) with "children" and "members" fields in the spec.  It also seems like this implies the need to support "dependencyOf" for other components here, perhaps with a third "dependencyOfComponents" field. Thoughts?

Otherwise, I'm not able to find any examples of non-defaulted kinds when calling `doEmit` in the existing codebase, nor any implementation that restricts the entity reference kinds that can be passed.  In fact, digging a little bit, it seems like each of these entity reference fields would reasonably accept any entity type passed, even if the reference didn't make sense.  Could anyone point me in the direction of this enforcement if it does exist, if we prefer this to be through a single field?

Reply:
We do also have the naming rules laid out next to the definitions of the relations: https://github.com/backstage/backstage/blob/c36a7847e49a1b61f76cff4451d30461c96e1246/packages/catalog-model/src/kinds/relations.ts#L18-L22

Reply:
And yes, just skip the restrictions in code. You can just emit what the component spec declares.

Reply:
Sounds good, I'll proceed without the restrictions in code and keep things as using the "dependsOn" and "dependencyOf" semantics.  Thanks!